File: src/prefix-sum/303.cpp
class NumArray
{
public:
  NumArray(vector<int> &nums)
  {
    prefixSums.push_back(0);

    for (int i = 0; i < nums.size(); i++)
    {
      prefixSums.push_back(prefixSums[i] + nums[i]);
    }
  }

  int sumRange(int left, int right)
  {
    return prefixSums[right + 1] - prefixSums[left];
  }

private:
  vector<int> prefixSums; // prefixSums[i] 表示小于i的所有数据的和
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray* obj = new NumArray(nums);
 * int param_1 = obj->sumRange(left,right);
 */

File: src/array/303.js
/**
 * @param {number[]} nums
 */
var NumArray = function (nums) {
  this.dp = []; // 从0到i-1的和

  this.dp[0] = 0;
  for (let i = 1; i <= nums.length; i++) {
    this.dp[i] = this.dp[i - 1] + nums[i - 1];
  }
};

/**
 * @param {number} left
 * @param {number} right
 * @return {number}
 */
NumArray.prototype.sumRange = function (left, right) {
  return this.dp[right + 1] - this.dp[left];
};

/**
 * Your NumArray object will be instantiated and called as such:
 * var obj = new NumArray(nums)
 * var param_1 = obj.sumRange(left,right)
 */
